pennylane/wires.py

```diff
@@ -0,0 +1,322 @@
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`Wires` class, which takes care of wire bookkeeping.
+"""
+from collections import Sequence, Iterable
+import numpy as np  # for random functions
+from copy import deepcopy
+
+# tolerance for integer-like wire indices
+TOLERANCE = 1e-8
+
+
+class WireError(Exception):
+    """Exception raised by a :class:`~.pennylane.wires.Wire` object when it is unable to process wires.
+    """
+
+
+def _clean(iterable):
+    """Converts 'iterable' into list of integers, and checks that all integers are unique and non-negative.
+
+        Args:
+            iterable (Iterable): some iterable object that represents a sequence of wire indices
+
+        Returns:
+            list[int]: cleaned wire index sequence
+
+        Raises:
+            WireError if elements cannot be turned into integers, if they are negative or not unique.
+        """
+
+    iterable = list(iterable)
+
+    # Turn elements into integers, if possible
+    for idx, w in enumerate(iterable):
+
+        if not isinstance(w, int):
+            # Try to parse integer-like elements to int
+            try:
+                # This works for floats and numpy.int
+                difference = abs(w - round(w))
+                if difference < TOLERANCE:
+                    iterable[idx] = int(w)
+                else:
+                    raise TypeError
+            except TypeError:
+                raise WireError(
+                    "Wire indices must be integers; got {} of type {}.".format(w, type(w))
+                )
+
+    # Check that indices are non-negative
+    for w in iterable:
+        if w < 0:
+            raise WireError("Wire indices must be non-negative; got index {}.".format(w))
+
+    # Check that indices are unique
+    if len(set(iterable)) != len(iterable):
+        raise WireError("Each wire must be represented by a unique index; got {}.".format(iterable))
+
+    return iterable
+
+
+class Wires(Sequence):
+    """
+    A bookkeeping class for wires, which are ordered collections of unique non-negative integers that
+    represent the index of a quantum subsystem such as a qubit or qmode.
+
+    Args:
+         wires (int or iterable): Iterable representing an ordered collection of unique wire indices.
+            The iterable can be of any common type such as list, tuple, range or numpy array.
+            The elements of the iterable must be non-negative integers. If elements are floats,
+            they are internally converted to integers, throwing an error if the rounding error exceeds TOLERANCE.
```

Comment:
If TOLERANCE is meant to be a module constant, the value should be included here rather than the constant name.